File: Cpp/0139_word_break.cpp
// https://leetcode.cn/problems/word-break/

class Solution {
public:
    bool wordBreak(string s, vector<string>& wordDict) {
        backtrack(s, wordDict, 0);
        return res;
    }

private:
    void backtrack(string s, vector<string>& wordDict, int idx) {
        if (idx == s.size()) {
            res = true;
            return;
        }

        for (string word: wordDict) {
            if (res) {
                return;
            }
            
            int length = word.size();
            string sub = s.substr(0, idx+length);
            if ((!word.compare(s.substr(idx, length))) &&
                (find(memory.begin(), memory.end(), sub) == memory.end())) {
                memory.push_back(sub);
                backtrack(s, wordDict, idx+length);
            }
        }
    }

    bool res = false;

    vector<string> memory;
};

File: Python3/0139_word_break.py
# https://leetcode.cn/problems/word-break/

class Solution:
    def wordBreak(self, s, wordDict):
        """
        Inputs:
            s: str
            wordDict: list[str]
        Outputs:
            res: bool
        """

        self.res = False
        self.memory = []
        self.backtrack(s, wordDict, 0)
        return self.res

    def backtrack(self, s, wordDict, idx):
        if idx == len(s):
            self.res = True
            return

        for word in wordDict:
            if self.res:
                return

            length = len(word)
            substr = s[0: idx+length]
            if (s[idx: idx+length] == word) and (substr not in self.memory):
                self.memory.append(substr)
                self.backtrack(s, wordDict, idx+length)
